lxd/apparmor/qemuimg.go

```diff
@@ -0,0 +1,175 @@
+package apparmor
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"text/template"
+
+	"github.com/lxc/lxd/lxd/sys"
+	"github.com/lxc/lxd/shared"
+	"github.com/lxc/lxd/shared/subprocess"
+)
+
+var qemuImgProfileTpl = template.Must(template.New("qemuImgProfile").Parse(`#include <tunables/global>
+profile "{{ .name }}" flags=(attach_disconnected,mediate_deleted) {
+  #include <abstractions/base>
+
+  capability dac_override,
+  capability dac_read_search,
+
+  {{range $index, $element := .allowedCmdPaths}}
+  	{{$element}} mixr,
+  {{- end }}
+
+  {{ .pathToImg }} rk,
+
+  {{- if .dstPath }}
+    {{ .dstPath }} rwk,
+  {{- end }}
+}
+`))
+
+type nullWriteCloser struct {
+	*bytes.Buffer
+}
+
+func (nwc *nullWriteCloser) Close() error {
+	return nil
+}
+
+func RunQemuImgWithApparmor(sysOS *sys.OS, cmd []string, imgPath string, dstPath string) (string, error) {
+	output, err := runQemuImgWithApparmor(cmd, nil, sysOS, imgPath, dstPath)
+	return output, err
+}
+
+func runQemuImgWithApparmor(cmd []string, stdin io.ReadCloser, sysOS *sys.OS, imgPath string, dstPath string) (string, error) {
```

Comment:
As this function is specific to `qemu-img` I think we should move the prlimit and qemu-img logic inside this function (along with the comment) and then have a struct called `QemuImgOpts` which can be used to pass the run-time invocation options (such as prlimit settings and extraction options).